vcomponent/src/vcomponent_hdmi_cec_device.c

```diff
@@ -0,0 +1,465 @@
+ * limitations under the License.
+*/
+
+#include <string.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <assert.h>
+
+#include "vcomponent_hdmi_cec_device.h"
+
+
+const static strVal_t gDIStrVal [] = {
+  { "TV", (int) DEVICE_TYPE_TV },
+  { "PlaybackDevice", (int) DEVICE_TYPE_PLAYBACK },
+  { "AudioSystem" , (int) DEVICE_TYPE_AUDIO_SYSTEM },
+  { "RecordingDevice", (int) DEVICE_TYPE_RECORDER },
+  { "Tuner", (int) DEVICE_TYPE_TUNER },
+  { "Reserved",  (int) DEVICE_TYPE_RESERVED },
+  { "FreeUse", (int) DEVICE_TYPE_FREEUSE },
+  { "Unregistered", (int) DEVICE_TYPE_UNREGISTERED },
+  { "Unknown", (int) DEVICE_TYPE_UNKNOWN }
+};
+
+const static strVal_t gVCStrVal [] = {
+  {"TOSHIBA", VENDOR_CODE_TOSHIBA},
+  {"SAMSUNG", VENDOR_CODE_SAMSUNG},
+  {"DENON", VENDOR_CODE_DENON},
+  {"MARANTZ", VENDOR_CODE_MARANTZ},
+  {"LOEWE", VENDOR_CODE_LOEWE},
+  {"ONKYO", VENDOR_CODE_ONKYO},
+  {"MEDION", VENDOR_CODE_MEDION},
+  {"TOSHIBA2", VENDOR_CODE_TOSHIBA},
+  {"APPLE", VENDOR_CODE_APPLE},
+  {"HARMAN_KARDON2", VENDOR_CODE_HARMAN_KARDON2},
+  {"GOOGLE", VENDOR_CODE_GOOGLE},
+  {"AKAI", VENDOR_CODE_AKAI},
+  {"AOC", VENDOR_CODE_AOC},
+  {"PANASONIC", VENDOR_CODE_PANASONIC},
+  {"PHILIPS", VENDOR_CODE_PHILIPS},
+  {"DAEWOO", VENDOR_CODE_DAEWOO},
+  {"YAMAHA", VENDOR_CODE_YAMAHA},
+  {"GRUNDIG", VENDOR_CODE_GRUNDIG},
+  {"PIONEER", VENDOR_CODE_PIONEER},
+  {"LG", VENDOR_CODE_LG},
+  {"SHARP", VENDOR_CODE_SHARP},
+  {"SONY", VENDOR_CODE_SONY},
+  {"BROADCOM", VENDOR_CODE_BROADCOM},
+  {"SHARP2", VENDOR_CODE_SHARP2},
+  {"VIZIO", VENDOR_CODE_VIZIO},
+  {"BENQ", VENDOR_CODE_BENQ},
+  {"HARMAN_KARDON", VENDOR_CODE_HARMAN_KARDON},
+  {"UNKNOWN", VENDOR_CODE_UNKNOWN},
+  };
+
+const static strVal_t gPSStrVal [] = {
+  { "on", (int)POWER_STATUS_ON  },
+  { "off", (int)POWER_STATUS_OFF  },
+  { "standby", (int)POWER_STATUS_STANDBY },
+  { "unknown", (int)POWER_STATUS_UNKNOWN }
+};
+
+/* Load the device info into the passed in vCHdmiCec_device_info_t*
+* prefix can be 
+*   "hdmicec/device_map/0"
+*   "hdmicec/device_map/0/children/0"
+*   "hdmicec/device_map/0/children/1"
+*   "hdmicec/device_map/0/children/1/children/0"
+*/
+void LoadDeviceInfo (ut_kvp_instance_t* instance, char* prefix, struct vCHdmiCec_device_info_t* device)
+{
+  char tmp[strlen(prefix) + 64];
+  char type[32];
+
+  if (device == NULL || instance == NULL || prefix == NULL)
+  {
+    assert(device != NULL);
+    assert(instance != NULL);
+    assert(prefix != NULL);
+    return;
+  }
+  strcpy(tmp, prefix);
+
+  strcpy(tmp + strlen(prefix), "/name");
+  ut_kvp_getStringField(instance, tmp, device->osd_name, MAX_OSD_NAME_LENGTH);
+
+  strcpy(tmp + strlen(prefix), "/active_source");
+  device->active_source = ut_kvp_getBoolField(instance, tmp);
+
+  strcpy(tmp + strlen(prefix), "/pwr_status");
+  ut_kvp_getStringField(instance, tmp, type, sizeof(type));
+  device->power_status = vCHdmiCec_GetValue(gPSStrVal, sizeof(gPSStrVal)/sizeof(strVal_t), type, (int)POWER_STATUS_UNKNOWN);
```

Comment:
`sizeof(gPSStrVal)/sizeof(strVal_t)`, you should make #define so you can remove errors from repeats where it can be incorrect.